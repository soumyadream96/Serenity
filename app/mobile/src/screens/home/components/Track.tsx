import React from 'react';
import { StyleSheet, TouchableOpacity } from 'react-native';
import FastImage from 'react-native-fast-image';
import { SongProps } from '@serenity/core';
import { DefaultImage } from '../../../components/DefaultImage';
import { Image, Text } from '@serenity/components';
import { Neomorph } from 'react-native-neomorph-shadows';
import { useTheme } from 'react-native-paper';

export interface TrackProps {
    track: SongProps;
    onPress: (track: SongProps) => void;
}

export function Track({ track, onPress }: TrackProps) {
    return (
        <TouchableOpacity
            style={[styles.item]}
            onPress={() => onPress(track)}
        >
            {/* <Neomorph
                // swapShadows // <- change zIndex of each shadow color
                darkShadowColor={'black'}
                lightShadowColor={'white'}
                style={{
                    shadowRadius: 6,
                    borderRadius: 12,
                    backgroundColor: colors.surface,
                    width: 120,
                    height: 120,
                    justifyContent: "center",
                    alignItems: "center",
                    //   display: "flex"
                }}
            >
                {track?.cover ? (
                    <FastImage
                        source={{
                            uri: track.cover,
                        }}
                        style={[styles.photo]}
                    />
                ) : (
                    <DefaultImage style={styles.photo} />
                )}
            </Neomorph> */}

            <Image source={track.cover} style={styles.photo}/>
            <Text numberOfLines={2} style={styles.title}>
                {track?.title}
            </Text>
        </TouchableOpacity>
    );
}

const styles = StyleSheet.create({
    item: {
        alignItems: 'center',
        marginBottom: 4,
        marginTop: 14,
        marginLeft: 12,
        width: 120,
    },
    title: {
        fontSize: 12,
        marginTop: 8,
        padding: 0,
        fontFamily: 'Nunito-Bold',
        includeFontPadding: false,
    },
    photo: {
        borderRadius: 12,
        // elevation: 4,
        height: 120,
        width: 120,
        backgroundColor: 'transparent'
    },
});


