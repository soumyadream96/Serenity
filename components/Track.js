import React, { Component } from 'react';
import { withTheme, Surface, List } from 'react-native-paper';
import { StyleSheet, NativeModules, LayoutAnimation } from 'react-native';
import { connect } from 'react-redux';
import _ from 'lodash'
import FastImage from 'react-native-fast-image'

import SwiperContainer from '../containers/SwiperContainer';
import { playMedia, addToQueue } from '../actions';
import Love from './Love';


const { UIManager } = NativeModules;

UIManager.setLayoutAnimationEnabledExperimental &&
    UIManager.setLayoutAnimationEnabledExperimental(true);

class Track extends Component {
    state = {
        hide: false
    }

    renderRightIcon = (props) => {
        const {
            track,
            active
        } = this.props;

        if (_.isEqual(active, track)) {
            const { colors } = this.props.theme;
            return <List.Icon {...props} icon="equalizer" color={colors.accent} />
        }
        return <Love {...props} />

    }
    // Playing a song is the song is not playing 
    play = () => {
        const {
            track,
            active
        } = this.props;
        if(active){
            if (_.isEqual(active, track)) {
                return false
            }
        }
        this.props.addToQueue(track);
        this.props.playMedia(track);
    }

    /*
        * Animation to remove an item from the list
        * may not be required for all render 
        * Adding duration would enhance the user experience
        * Testing has to be done
    */ 
    close = () => {
        LayoutAnimation.spring();
        this.setState({ hide: true })
    }

    render() {
        const {
            track
        } = this.props;

        if (this.state.hide && track) {
            return false
        }
        return (
            <SwiperContainer close={() => this.close()}>
                <Surface style={styles.surface}>
                    <List.Item
                        item={track}
                        title={track.title}
                        description={track.artist}
                        left={props => (
                            <FastImage {...props} source={{ uri: track.artwork }} style={styles.icons} />
                        )}
                        right={props => this.renderRightIcon(props)}
                        onPress={() => this.play()}
                    />
                </Surface>
            </SwiperContainer>
        );
    }
}

const mapStateToProps = state => ({
    active: state.media.active,
});

export default connect(mapStateToProps, { playMedia, addToQueue })(withTheme(Track));

const styles = StyleSheet.create(
    {
        icons: {
            width: 50,
            borderRadius: 4
        },
        surface: {
            padding: 0,
            margin: 0,
            borderRadius: 4
        }
    });