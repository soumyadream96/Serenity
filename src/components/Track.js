import React from 'react';
import { withTheme, List } from 'react-native-paper';
import { StyleSheet, View } from 'react-native';
import PropTypes from 'prop-types';

/*
    TODO: 
    - may not be required for all render 
    - Adding duration would enhance the user experience
    - Testing has to be done
*/

// FIXME: Testing the application
const Track = ({ theme: { colors }, track, active, play }) => (
  <View style={[styles.surface, { backgroundColor: colors.background }]}>
    <List.Item
      title={track.title}
      description={track.artist ? track.artist : track.album}
      right={props =>
        active ? (
          <List.Icon {...props} icon="poll" color={colors.accent} />
        ) : (
          // <List.Icon {...props} icon="more-vert" onPress={() => bs.current.snapTo(1)}/>
          false
        )
      }
      onPress={() => play()}
    />
  </View>
);

Track.propTypes = {
  track: PropTypes.object.isRequired,
  active: PropTypes.bool.isRequired,
};

export default withTheme(Track);

const styles = StyleSheet.create({
  surface: {
    padding: 0,
    margin: 0,
    borderRadius: 4,
  },
});
