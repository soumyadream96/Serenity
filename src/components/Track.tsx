import React from 'react';
import { withTheme, List, Theme } from 'react-native-paper';
import { StyleSheet, View } from 'react-native';

interface TrackProps {
  title: string;
  album?: string;
  artist?: string;
}

interface Props {
  theme: Theme;
  track: TrackProps;
  active: boolean;
  play(): void;
}

const Track = ({ theme: { colors }, track, active, play }: Props) => (
  <View style={[styles.surface, { backgroundColor: colors.background }]}>
    <List.Item
      title={track.title}
      description={track.artist ? track.artist : track.album}
      right={props =>
        active ? (
          <List.Icon {...props} icon="poll" color={colors.accent} />
        ) : (
          // <List.Icon {...props} icon="more-vert" onPress={() => bs.current.snapTo(1)}/>
          false
        )
      }
      onPress={() => play()}
    />
  </View>
);

export default withTheme(Track);

const styles = StyleSheet.create({
  surface: {
    padding: 0,
    margin: 0,
    borderRadius: 4,
  },
});
