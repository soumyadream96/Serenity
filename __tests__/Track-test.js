/**
 * @format
 */

import 'react-native';
import React from 'react';
import Track from '../src/components/Track';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

it('In active track renders correctly', () => {
  const track = {title: 'title', artist: 'Tester'};
  const play = jest.fn();
  renderer.create(
    <Track track={track} active={false} play={play} />,
  );
});

it('Active track renders correctly', () => {
  const track = {title: 'title', artist: 'Tester'};
  const play = jest.fn();
  renderer.create(
    <Track track={track} active={true} play={play} />,
  );
});

it('Track with album name renders correctly', () => {
  const track = {title: 'title', album: 'Testing'};
  const play = jest.fn();
  renderer.create(
    <Track track={track} active={true} play={play} />,
  );
});
