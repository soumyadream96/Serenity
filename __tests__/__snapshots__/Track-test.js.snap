// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blank component renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 4,
        "margin": 0,
        "padding": 0,
      },
      Object {
        "backgroundColor": "#f6f6f6",
      },
    ]
  }
>
  <View
    accessible={true}
    clickable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        false,
        Array [
          Object {
            "padding": 8,
          },
          undefined,
        ],
      ]
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <View
        pointerEvents="none"
        style={
          Array [
            Object {
              "margin": 8,
            },
            Object {
              "flex": 1,
              "justifyContent": "center",
            },
          ]
        }
      >
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "Helvetica Neue",
                "textAlign": "left",
                "writingDirection": "ltr",
              },
              Array [
                Object {
                  "fontSize": 16,
                },
                Object {
                  "color": "rgba(0, 0, 0, 0.87)",
                },
                undefined,
              ],
            ]
          }
        >
          title
        </Text>
        <Text
          numberOfLines={2}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "Helvetica Neue",
                "textAlign": "left",
                "writingDirection": "ltr",
              },
              Array [
                Object {
                  "fontSize": 14,
                },
                Object {
                  "color": "rgba(0, 0, 0, 0.54)",
                },
                undefined,
              ],
            ]
          }
        >
          Testing
        </Text>
      </View>
      <View
        pointerEvents="box-none"
        style={
          Array [
            Object {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "margin": 8,
              "width": 40,
            },
            undefined,
          ]
        }
      >
        <Text
          accessibilityElementsHidden={true}
          allowFontScaling={false}
          importantForAccessibility="no-hide-descendants"
          pointerEvents="none"
          style={
            Array [
              Object {
                "color": "#03dac4",
                "fontSize": 24,
              },
              Array [
                Object {
                  "transform": Array [
                    Object {
                      "scaleX": 1,
                    },
                  ],
                },
                Object {
                  "backgroundColor": "transparent",
                },
              ],
              Object {
                "fontFamily": "Material Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          ÓÄù
        </Text>
      </View>
    </View>
  </View>
</View>
`;
